<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2//EN">
<HTML>
<HEAD>
	<META HTTP-EQUIV="CONTENT-TYPE" CONTENT="text/html; charset=iso-8859-1">
	<TITLE>BAMBUS 2.3 User Manual</TITLE>
	<META NAME="GENERATOR" CONTENT="OpenOffice.org 1.0.1  (Linux)">
	<META NAME="CREATED" CONTENT="20021028;15303200">
	<META NAME="CHANGEDBY" CONTENT="Mihai Pop">
	<META NAME="CHANGED" CONTENT="20021112;11554600">
	<STYLE>
	<!--
		TD P { color: #000000 }
		P { color: #000000 }
		H3 { color: #ff0000 }
		H3.western { font-style: italic }
	-->
	</STYLE>
</HEAD>
<BODY LANG="en-US" TEXT="#000000" BGCOLOR="#ffffff">
<TABLE WIDTH=100% BORDER=0 CELLPADDING=2 CELLSPACING=0 STYLE="page-break-before: always"> 
<COL WIDTH=100%> <TR> <TD WIDTH=100%> <TABLE ALIGN=LEFT WIDTH=80% BORDER=0 CELLPADDING=5 CELLSPACING=5> 
<TR> <TD WIDTH=23% HEIGHT=108> <P><IMG SRC="Manual_files/bambus.jpg" NAME="Graphic1" ALIGN=BOTTOM WIDTH=79 HEIGHT=78 BORDER=0></P></TD><TD WIDTH=77%> 
<H3 CLASS="western"><FONT COLOR="#cc0000"><B>BAMBUS 2.33 User Manual</B></FONT></H3></TD></TR> 
</TABLE></TD></TR> <TR> <TD WIDTH=100%> <H3 CLASS="western"><FONT COLOR="#ff0000"><I>Index</I></FONT></H3><OL> 
<LI> <A HREF="#intro">Introduction</A> <ul> <li><A HREF="#algorithm">Algorithm 
overview</A></li></ul><li> <A HREF="#install">Installation</A></li><OL> <LI> <A HREF="#prereqs">Prerequisites</A> 
<LI> <A HREF="#download">Obtaining BAMBUS</A> <LI> <A HREF="#doit">Actual installation</A> 
</OL><LI> <A HREF="#run">Running BAMBUS</A> <OL> <LI> <A HREF="#options">Command 
line options</A> <OL> <LI> <A HREF="#confopt">Configuration options</A> <LI> <A HREF="#inputopt">Input 
options</A> <LI> <A HREF="#mateopt">Mate-pair information</A> <LI> <A HREF="#outputopt">Output 
options</A> <LI> <A HREF="#repopt">Repeat screening</A> <LI><A HREF="#flow">Flow 
Control</A></OL><LI> <A HREF="#usecase">Typical use cases</A> <OL> <LI> <A HREF="#asmout">Using 
assembler output</A> <LI> <A HREF="#infoout">Getting more (or less) information 
from the output</A> <LI> <A HREF="#xmlinfo">Adding additional XML linking information</A> 
<LI> <A HREF="#muminfo">Using MUMmer links</A> <LI> <A HREF="#makefiles">Using 
Makefiles</A> <LI><A HREF="#untangle">Untangling scaffolds</A></OL></OL><LI> <A HREF="#formats">File 
formats</A> <OL> <LI><A HREF="#conffile">The configuration file</A> <LI> <A HREF="#matesfile">The 
.mates file</A> <LI> <A HREF="#repfile">Repeat files</A> <LI> <A HREF="#EVXML">.evidence.xml</A> 
<LI> <A HREF="#outxml">.out.xml</A> <LI> <A HREF="#summary">.stats file</A> <LI> 
<A HREF="#details">.details file</A> <LI> <A HREF="#dotfile">.dot file</A> <LI> 
<A HREF="#psfile">.ps file</A> <LI><A HREF="#oo">.oo file</A><LI><A HREF="#sum">.sum 
file</A></OL><LI> <A HREF="#problems">Known problems</A> <LI> <A HREF="#contact">Contact 
information</A> </OL><HR width="100%"> <H3 CLASS="western"><a name="intro"></a>1. 
Introduction</H3><P>The output of most shotgun sequence assembly programs (such 
as TIGR Assembler, phrap, or CAP3) consists in a set of un-related contigs, whose 
order and orientation along the chromosome is unknown. Scaffolding represents 
the task of ordering and orienting these contigs by using additional information 
about their relative placement. Traditionally such information was identified 
from the pairing of reads from the opposite ends of an insert in double-barelled 
shotgun experiments (see figure). </P><P ALIGN=CENTER><IMG SRC="Manual_files/scaffold-www.gif" NAME="Scaffold" ALT="scaffold image" ALIGN=BOTTOM WIDTH=354 HEIGHT=116 BORDER=0></P><P ALIGN=LEFT>Besides 
clone mate information, homology data, physical maps, or gene synteny information 
can be used to derive relationships between contigs. BAMBUS is a program written 
to handle all such types of data in a generic fashion in order to build contig 
scaffolds. In order to account for the varied quality of linking information BAMBUS 
allows the user to specify a hierarchy that will consider the most reliable linking 
data first, then expand the scaffolds using less reliable data. This approach 
minimizes the effect of errors inherent to experimentally-derived linking information.</P><H3 CLASS="western" ALIGN=LEFT><A NAME="algorithm"></A><FONT COLOR="#ff0000"><I>Algorithm 
overview</I></FONT></H3><P ALIGN=LEFT>BAMBUS has two main modes of operation: 
hierarchical - when the links are considered in the order of their priorities, 
and standard - when all the links are considered at the same priority level. In 
the latter case, all links connecting two contigs are bundled together to form 
a &quot;gap&quot;, or &quot;edge&quot; if you look at the problem from a graph 
theoretical point of view. Each link is checked for validity before being added 
to a bundle. Thus, a link is considered invalid due to a length constraint if 
it forces the two contigs to overlap. In other words, if we assume the largest 
possible size for the insert corresponding to the link, if the coordinates of 
the mate-pair within the contigs force the contigs to overlap, we discard the 
link assuming it is due to a misassembly (Note: this behaviour can be specified 
on a link-by-link basis). Links that pass this test are further checked for consistency. 
If they do not all agree in the relative orientations of the contigs they imply, 
a majority rule is used to retain only those links that agree both in orientation 
and length. Furthermore, the bundle, or edge, is retained if it contains at least 
two links (this parameter, <FONT FACE="Courier New, Courier, mono">redundancy</FONT>, 
can actually be tuned by the user). After this step is complete, the resulting 
graph gets traversed twice, once to assign consistent orientations to all the 
contigs, and the second time to determine the order of the contigs along the chromosome. 
Note that inconsistencies in order are currently allowed since they can provide 
finishing teams with useful information. In a future version of BAMBUS we will 
add an option to allow the generation of unambiguous scaffolds as well. Currently 
you can generate such unambiguous scaffolds using the <FONT FACE="Courier New, Courier, monospace"><A HREF="#untangle">untangle</A></FONT> 
program. </P><H3 CLASS="western"><a name="install"></a>2. Installation</H3><H3 CLASS="western"><FONT COLOR="#ff0000"><I><a name="prereqs"></a>2.1. 
Prerequisites</I></FONT></H3><UL> <LI> <font color="#000000">Unix (tested on RedHat 
6.2 and higher, Solaris, and OSF 1)</font> <B>Does not work in cygwin!</B><LI> 
<FONT COLOR="#000000">Perl 5.6 or later</FONT> <LI> <FONT COLOR="#000000"><A HREF="http://www.cpan.org/">XML::Parser</A> 
perl module</FONT> <LI> <FONT COLOR="#000000"><A HREF="http://www.cpan.org/">Config::IniFiles</A> 
perl module</FONT> <LI> <P><FONT COLOR="#000000"><A HREF="http://www.research.att.com/sw/tools/graphviz/">GraphViz</A> 
package</FONT> </P></UL><H3 CLASS="western"><A NAME="download"></A><FONT COLOR="#ff0000"><I>2.2. 
Obtaining BAMBUS</I></FONT></H3><P>BAMBUS 2.3 is available free of charge under 
the open-source <A HREF="http://www.opensource.org/licenses/artistic-license.php">Artistic 
License</A><BR>. <A HREF="http://www.opensource.org/docs/definition.php"><IMG SRC="Manual_files/osi-certified-120x100.png" WIDTH="120" HEIGHT="100" BORDER="0"></A></P><P CLASS="western">To 
download Bambus please <A HREF="ftp://ftp.tigr.org/pub/software/BAMBUS">click 
here</A>.</P><P CLASS="western">If you wish to receive update notices you can 
join the BAMBUS mailing list by <A HREF="http://www.tigr.org/tigr-scripts/software/update.pl?program=BAMBUS">clicking 
here</A>.</P><H3 CLASS="western"><A NAME="doit"></A><FONT COLOR="#ff0000"><I>2.3. 
Actual installation</I></FONT></H3><OL> <LI> <FONT COLOR="#000000">Find a directory 
where you want the software installed. For example <FONT FACE="Courier New, Courier, mono">/users/home/CoolUser</FONT></FONT> 
<LI> <FONT COLOR="#000000">Place the tar file <FONT FACE="Courier New, Courier, mono">bambus-2.33.tar.gz</FONT> 
in this directory</FONT> <LI> <FONT COLOR="#000000">Unpack the distribution<BR> 
<FONT FACE="Courier New, Courier, mono">% tar xvzf bambus-2.33.tar.gz</FONT></FONT> 
<LI> <FONT COLOR="#000000">Go into the newly created directory<BR> <FONT FACE="Courier New, Courier, mono">% 
cd bambus-2.33</FONT></FONT><LI> <FONT COLOR="#000000">Carefully read the documentation</FONT> 
<LI> Edit the file Makefile and change the value of BASEDIR to the correct installation 
path, for example: <FONT FACE="Courier New"><BR>BASEDIR = /users/home/CoolUser</FONT> 
<LI> Also check if the path for perl (in variable PERL) matches the one on your 
system. Several common options are:<BR><FONT FACE="Courier New, Courier, monospace">PERL 
= /usr/local/bin/perl<BR>PERL = /usr/bin/perl</FONT><LI> Type <FONT FACE="Courier New">'gmake 
install' </FONT> to have all the software installed. The executables will be in 
/<FONT FACE="Courier New">users/home/CoolUser/bin/ </FONT>and the documentation 
in <FONT FACE="Courier New">/users/home/CoolUser/doc.</FONT> <LI> <P>To get started 
you can try out the small test dataset provided in data/test_run by typing:<FONT FACE="Courier New"><BR> 
goBambus -c test.contig -m test.mates -o test-bambus</FONT></P></OL><H3 CLASS="western"><A NAME="run"></A><FONT COLOR="#ff0000"><I><BR> 
3. Running BAMBUS</I></FONT></H3><P>The input to Bambus consists in a set of links 
between contigs. These links can be inferred from the pairing of reads belonging 
to the same insert, from physical map data, or from alignment to another genome. 
To accomodate these types of sources of linking information, and to allow for 
further extensions, the input to Bambus is presented in a general purpose XML 
format. For more information about this format click <A HREF="#EVXML">here.</A></P><P>The 
following sections describe how the XML file can be generated for each type of 
linking data.</P><P>Linking information is grouped into &quot;libraries&quot;. 
All links within a library have similar parameters. In the case of mate-pair linking 
information the libraries match the standard definition. For other linking data, 
for example MUMmer links, the library grouping reflects specific characteristics 
of the linking data. For example all MUMmer links are grouped within a library 
called &quot;MUMmer&quot;. </P><H3 CLASS="western"><A NAME="options"></A>3.1 Command 
line options</H3><P>Bambus consists in a collection of programs controlled by 
a script called <FONT FACE="Courier, monospace">goBambus.</FONT> The execution 
is controlled by the following sets of parameters. To obtain a list of all parameters 
and basic help information you need to type:</P><P><FONT FACE="Courier, monospace">goBambus 
-h</FONT><BR> or<BR> <FONT FACE="Courier, monospace">goBambus -help</FONT></P><H3 CLASS="western"><A NAME="confopt"></A>3.1.1 
Configuration options</H3><P><FONT FACE="Courier, monospace">-C &lt;conf_file&gt;<FONT FACE="Thorndale"> 
. This parameter specifies the configuration file used by Bambus. A configuration 
file is not required, however it is useful if you want to change the default parameters. 
Click <A HREF="#conffile">here</A> for a detailed description of the configuration 
information.</FONT></FONT></P><P><FONT FACE="Thorndale">Whenever you run Bambus 
it will generate a default configuration file called <FONT FACE="Courier New, Courier, monospace">default.conf</FONT>. 
If you want to modify the parameters, you will need to rename this file and then 
edit it.</FONT></P><H3 CLASS="western"><A NAME="inputopt"></A>3.1.2 Input Options</H3><P><FONT FACE="Thorndale"><FONT FACE="Courier, monospace">-a 
&lt;asm_file&gt; .</FONT> Use a TIGR Assembler .asm file as source of linking 
information. <BR> -<FONT FACE="Courier, monospace">c &lt;contig_file&gt; </FONT>. 
Use a GDE formatted <FONT FACE="Courier, monospace">.contig</FONT> assembly file 
as source of linking information. The .<FONT FACE="Courier, monospace">contig</FONT> 
files can be obtained either from TIGR Assembler output (by concatenating the 
contents of the <FONT FACE="Courier, monospace">.align</FONT> directory) or by 
converting <FONT FACE="Courier, monospace">.ace</FONT> files (created by phrap 
or Consed) using the <A HREF="ftp://ftp.tigr.org/pub/software/ta2ace">ta2ace</A> 
package.<BR> <FONT FACE="Courier, monospace">-x &lt;xml_file&gt;</FONT>. Use additional 
XML formatted linking information. You can use the <FONT FACE="Courier, monospace">-x</FONT> 
option multiple times. The format for the XML file is described <A HREF="#EVXML">here</A>.<BR> 
<FONT FACE="Courier, monospace">-mx &lt;make_file&gt;. </FONT>Use a gmake compatible 
make file to specify how the additional XML information is generated. Each of 
the files created must end in &quot;.<FONT FACE="Courier, monospace">xml</FONT>&quot;. 
Moreover, only the final XML targets are allowed to end in &quot;<FONT FACE="Courier, monospace">.xml</FONT>&quot; 
to prevent goBambus from getting confused. A more detailed description on using 
makefiles is presented <A HREF="#makefiles">here.</A></FONT></P><H3 CLASS="western"><A NAME="mateopt"></A>3.1.3 
Mate-pair Information</H3><P>Assembly output (such as .asm and .contig files) 
does not generally contain any information about the pairing of reads coming from 
opposite ends of the same insert, nor about the membership of inserts to libraries. 
The following parameters specify how to obtain this mate-pair information.</P><P><FONT FACE="Courier, monospace">-m 
&lt;mates_file&gt;</FONT> . The mates file allows you to specify both library 
information and mate pairing information, using a system based on Perl regular 
expressions. For more information about the file format click <A HREF="#matesfile">here.</A><BR> 
<FONT FACE="Courier, monospace">-D &lt;database&gt; </FONT>. At TIGR, instead 
of a mates file you can opt to use a database to obtain library and mate information. 
Note that the <FONT FACE="Courier, monospace">-m</FONT> and <FONT FACE="Courier, monospace">-D</FONT> 
options are complementary. The following three parameters specify database options:<BR> 
<FONT FACE="Courier, monospace">-S &lt;server&gt;</FONT> . Select a specific database 
server<BR> <FONT FACE="Courier, monospace">-U &lt;user&gt;</FONT> . Log in with 
the specified user name<BR> <FONT FACE="Courier, monospace">-P &lt;passwd&gt;</FONT> 
. Log in with the specified password</P><H3 CLASS="western"><A NAME="outputopt"></A>3.1.4 
Output Options</H3><P><FONT FACE="Courier, monospace">-o &lt;output_prefix&gt;</FONT> 
. All the output file names will be generated from the <FONT FACE="Courier, monospace">&lt;output_prefix&gt;</FONT> 
prefix.</P><H3 CLASS="western"><A NAME="repopt"></A>3.1.5 Repeat Screening</H3><P>Misassembled 
repeats can confuse Bambus, therefore it makes sense to screen the known repeats 
that may be misassembled. Bambus can use two types of repeat screening information:</P><P>-r 
. Screen against the repeats listed in the database (option -D must be provided)<BR> 
-r &lt;repfile&gt; . Screen against repeats described in &lt;repfile&gt;. For 
a detailed description of the repeat file format click <A HREF="#repfile">here.</A><BR></P><H3><I><A NAME="flow"></A>3.1.6 
Flow-control </I></H3><P><FONT FACE="Courier New, Courier, monospace">-start &lt;num&gt;<BR>-end 
&lt;num&gt;.</FONT> The execution of Bambus consists in a set of steps. The user 
can choose to start or end at a specific step for example in the case when she 
modifies the configuration file and wants to re-run just the final steps of the 
scaffolders. This feature should be used with caution. </P><H3 CLASS="western"><A NAME="usecase"></A>3.2 
Typical use cases</H3><H3 CLASS="western"><A NAME="asmout"></A>3.2.1 Using assembler 
output</H3><P>For convenience, Bambus can use the output of an assembly program 
and automatically generate the required linking information. Two types of information 
are required:</P><UL> <LI> <P>the tiling of reads in the assembly</P><LI> <P>information 
about the pairing of reads</P></UL><P>The first type of information can be obtained 
from the <FONT FACE="Courier New">.asm</FONT> or <FONT FACE="Courier New">.contig</FONT> 
output formats of TIGR Assembler. For more information about these file formats 
please refer to the TIGR Assembler <A HREF="http://www.tigr.org/software/assembler/helpfile.html">documentation. 
</A>Note that the .contig file may be stripped of all sequence data as the only 
information used is that contained in the lines starting with #. In case you are 
using an assembler other than TIGR Assembler you may need to use the <A HREF="ftp://ftp.tigr.org/pub/software/ta2ace">ta2ace</A> 
package to convert the output into the <FONT FACE="Courier New">.contig</FONT> 
format.</P><P>The second type of information can be obtained from a .mates file 
whose format is presented <A HREF="#matesfile">below,</A> or from the database 
in case you are at TIGR. </P><P>A typical command line for someone outside of 
TIGR is:</P><P><FONT FACE="Courier, monospace">goBambus -c test.contig -m test.mates 
-o test</FONT></P><P>and the execution will generate the following files:</P><P><FONT FACE="Courier, monospace">test.stats 
</FONT>- statistics on scaffolds and libraries (format described <A HREF="#summary">here</A>)<BR> 
<FONT FACE="Courier, monospace">test.details </FONT>- detailed information about 
each contig pair (also called &quot;gap&quot;). This file is described<A HREF="#details"> 
here</A>.<BR> <FONT FACE="Courier, monospace">test.dot </FONT>- GraphViz formatted 
description of the contig linkage information. This file is described <A HREF="#dotfile">here</A>.<BR> 
<FONT FACE="Courier, monospace">test.evidence.xml</FONT> - XML representation 
of all the linking evidence provided. The format is described <A HREF="#EVXML">here</A>.<BR> 
<FONT FACE="Courier, monospace">test.out.xml</FONT> - XML representation of the 
scaffolds. Format described <A HREF="#outxml">here</A>.<BR> <FONT FACE="Courier, monospace">test.lib</FONT> 
- list of the codes associated with each input library. These codes are reported 
on the links in the <A HREF="#psfile">output file</A>.<BR><FONT FACE="Courier New, Courier, monospace">test.sum</FONT> 
- one-line sumaries of all scaffolds. Format described <A HREF="#sum">here</A>.<BR><FONT FACE="Courier New, Courier, monospace">test.oo</FONT> 
- order and orientation information for all the contigs. Format described <A HREF="#oo">here</A>.</P><P>Additionally, 
the program generates some &quot;working&quot; files that are currently not removed 
as they provide useful debugging information.</P><P><FONT FACE="Courier, monospace">test.detective.xml</FONT> 
- linking information derived from assembler input alone. It's the same as test.evidence.xml 
unless additional XML files are provided in the input.<BR> <FONT FACE="Courier, monospace">test.inp</FONT> 
- input to the core scaffolding engine.<BR> <FONT FACE="Courier, monospace">test.grommit.conf</FONT> 
- configuration information for the scaffolding engine<BR> </P><P>To view the 
graphical output you need to convert it to Postscript with the command:</P><P><FONT FACE="Courier, monospace">dot 
-Tps -o test.ps test.dot</FONT></P><P>Then you can view the postscript file with 
the command:</P><P><FONT FACE="Courier, monospace">gv test.ps</FONT></P><P>At 
TIGR, instead of specifying a <FONT FACE="Courier, monospace">.mates </FONT>file 
you can use the database:</P><P><FONT FACE="Courier, monospace">goBambus -c test.contig 
-D gbx -U access -P access -o test</FONT></P><H3 CLASS="western"><A NAME="infoout"></A>3.2.2 
Getting more (or less) information from the output</H3><P>You can modify the information 
you get in the output by directly calling the <FONT FACE="Courier, monospace">printScaff 
</FONT>command. PrintScaff requires as inputs the <FONT FACE="Courier, monospace">.evidence.xml</FONT> 
file, the <FONT FACE="Courier, monospace">.out.xml</FONT> file and the <FONT FACE="Courier, monospace">.lib</FONT> 
files produce by Bambus, the minimal invocation being:</P><P><FONT FACE="Courier, monospace">printScaff 
-e test.evidence.xml -s test.out.xml -l test.lib</FONT></P><P>Optionally you may 
specify the output prefix with option <FONT FACE="Courier, monospace">-o</FONT> 
(just like in the Bambus invocation).</P><P>PrintScaff has parameters that control 
what gets reported, and in what format. In the first category you have:</P><P><FONT FACE="Courier, monospace">-dot</FONT> 
- produce a <FONT FACE="Courier, monospace">.dot</FONT> file<BR> <FONT FACE="Courier, monospace">-detail</FONT> 
- produce a <FONT FACE="Courier, monospace">.details</FONT> file<br> <font face="Courier New, Courier, monospace">-oo</font> 
- produce a <A HREF="#oo">.oo</A> file listing all the contigs in each scaffold<br> 
<font face="Courier New, Courier, monospace">-sum</font> - produce a <A HREF="#sum">.sum</A> 
tab delimited list of scaffold stats (#contigs, size, and span)<br> <font face="Courier New, Courier, monospace">-f 
&lt;fasta_file&gt;</font> - generate a pseudo-molecule for each scaffold using 
the contig sequences listed in <font face="Courier New, Courier, monospace">&lt;fasta_file&gt;.</font><br> 
&nbsp;&nbsp;note that the contigs in the scaffold files are names &quot;<font face="Courier New, Courier, monospace">contig_&lt;id&gt;</font>&quot; 
while the contigs in the <font face="Courier New, Courier, monospace">&lt;fasta_file&gt;</font> 
must simply be called &quot;<font face="Courier New, Courier, monospace">&lt;id&gt;</font>&quot;. 
Note that this option together with the default <font face="Courier New, Courier, monospace">-merge</font> 
option (see below) is meaningless unless you've <a href="#untangle">untangled</a> 
the scaffold.<BR> <FONT FACE="Courier, monospace">-phys</FONT> - create a <FONT FACE="Courier, monospace">.phys</FONT> 
file listing all &quot;gaps&quot; spanned only by the specified libraries:</P><P><FONT FACE="Courier, monospace">printScaff 
-e test.evidence.xml -s test.out.xml -l test.lib -phys MUMmer</FONT></P><P>will 
generate a file called <FONT FACE="Courier, monospace">test.phys</FONT> that contains 
all contig pairs linked by nothing but MUMmer links.</P><P>In the second category 
you have:</P><P><FONT FACE="Courier, monospace">-page</FONT> - produce <FONT FACE="Courier, monospace">.dot</FONT> 
file formatted for printing on 8.5x11&quot; paper<BR> <FONT FACE="Courier, monospace">-plot</FONT> 
- produce .<FONT FACE="Courier, monospace">dot</FONT> file formatted for printing 
on a plotter (36x48&quot; paper)<BR> <FONT FACE="Courier, monospace">-unused</FONT> 
- draw edges corresponding to unused links</P><P><font face="Courier New, Courier, monospace">-merge</font> 
- (default) when the <font face="Courier New, Courier, monospace">-f</font> option 
is given, produce a pseudo-molecule for each scaffold by adding 60 <font face="Courier New, Courier, monospace">N</font> 
characters between the contigs in the scaffold<br> <font face="Courier New, Courier, monospace">-nomerge</font> 
- when the <font face="Courier New, Courier, monospace">-f </font>option is given, 
create a fasta file containing all contigs in each scaffold in the correct orientation</P><H3 CLASS="western"><A NAME="xmlinfo"></A>3.2.3 
Adding additional XML linking information</H3><P>You can easily add additional 
linking information as long as it's in a format similar to the<A HREF="#EVXML"><FONT FACE="Courier, monospace"> 
.evidence.xml </FONT></A>file. Assuming we have two such files &quot;link1.xml&quot; 
and &quot;link2.xml&quot; you can call Bambus as follows:</P><P><FONT FACE="Courier, monospace">goBambus 
-c test.contig -m test.mates -x link1.xml -x link2.xml -o test</FONT></P><P>The 
result being the same as if you concatenated the XML file inferred from the .contig 
file together with the two additional XML files.</P><H3 CLASS="western"><A NAME="muminfo"></A>3.2.4 
Using MUMmer links</H3><P>In case you are sequencing a genome for which a close 
relative has been completed, you can use <A HREF="http://www.tigr.org/software/mummer">MUMmer</A> 
to infer links between contigs. Assume you have the reference genome in a file 
&quot;<FONT FACE="Courier, monospace">ref.fasta</FONT>&quot; and all the contigs 
from the genome you are assembling in a file called &quot;<FONT FACE="Courier, monospace">test.fasta</FONT>&quot;, 
you can run MUMmer to align the test genome to the reference:</P><P><FONT FACE="Courier, monospace">nucmer 
-a max-match ref.fasta test.fasta </FONT> </P><P>The output will be placed in 
a file called <FONT FACE="Courier, monospace">out.delta</FONT>. You can then convert 
the alignment information into an XML file suitable for Bambus as follows:</P><P><FONT FACE="Courier, monospace">show-tiling 
-x out.delta &gt; test.mum.xml</FONT></P><P>And finally you can provide this information 
to Bambus:</P><P><FONT FACE="Courier, monospace">goBambus -c test.contig -m test.mates 
-x test.mum.xml -o test</FONT></P><H3 CLASS="western"><A NAME="makefiles"></A> 
3.2.5 Using Makefiles</H3><P>If you have multiple XML files, it becomes tedious 
to generate each file separately and then provide them to Bambus with multiple 
-x options. You can create a Makefile compatible with <A HREF="http://www.gnu.org/software/make/make.html">GNU 
make </A>that generates each of the XML files. You must be careful that the targets 
for all the XML files end in &quot;<FONT FACE="Courier, monospace">.xml</FONT>&quot;, 
and no other targets (such as intermediate targets) have the same suffix. Bambus 
will assume that all .<FONT FACE="Courier, monospace">xml</FONT> files produced 
by the Makefile must be added to the input.</P><P>An example of a simple Makefile 
that can generate the mummer information described above is:</P><P><FONT FACE="Courier, monospace">test.mum.xml: 
ref.fasta test.fasta<BR> nucmer -a max-match ref.fasta test.fasta<BR> show-tiling 
-x out.delta &gt; test.mum.xml<BR> </FONT></P><h3><BR> <a name="untangle"></a>3.2.6 
Untangling scaffolds</h3><P>By default, the scaffolds resulting from Bambus are 
potentially ambiguous as two or more contigs may occupy the same place in the 
genome. Such situations occur either due to misassembled repeats, or when assembling 
different haplotypes. Bambus contains a utility that does a best effort attempt 
to disambiguate the scaffolds, by breaking them into a set of non-ambiguous scaffolds. 
The algorithm is greedy and does not guarantee that an optimal solution (e.g. 
a minimum number of longest scaffolds) is produced. </P><P>To run the untangler 
type:</P><P><font face="Courier New, Courier, monospace">untangle -e test.evidence.xml 
-s test.out.xml -o test.untangle.xml</font></P><P>The test.untangle.xml file has 
the same format as the output file (<a href="#outxml"><font face="Courier New, Courier, monospace">.out.xml</font></a>) 
and for each scaffold (e.g. <font face="Courier New, Courier, monospace">scaff_1</font>) 
contains one or more unambiguous scaffolds called &quot;<font face="Courier New, Courier, monospace">scaff_1_1, 
scaff_1_2</font>, etc&quot;. You can use the <font face="Courier New, Courier, monospace"><a href="#infoout">printScaff</a> 
</font>command to process the untangled file the same way you would process a 
normal Bambus output.<BR> </P><H3 CLASS="western"><A NAME="formats"></A>4. File 
formats</H3><H3 CLASS="western"><FONT COLOR="#ff0000"><a name="conffile"></a>4.1 
The configuration file</FONT></H3><P>Scaffolding parameters can be specified in 
a configuration file. Whenever you run Bambus a default configuration file is 
generated as <FONT FACE="Courier New, Courier, monospace">default.conf. </FONT>To 
create your own file you should rename this file then edit it. The following types 
of information can be specified:</P><UL> <LI> <P><B>Comments.</B> Lines starting 
with # are considered comments and are ignored by the software.</P><LI> <P><B>Priority 
information.</B> Specifies an order in which the libraries are considered. For 
example, the line:<BR> <FONT FACE="Courier, monospace">priority MUMmer 2</FONT><BR> 
specifies priority 2 for all MUMmer links. The libraries are considered in the 
increasing order of priorities. <BR> <BR> If no configuration file is provided, 
all libraries are assumed to have the same priority. If a configuration file is 
provided, only those libraries that have a priority record are used in scaffolding. 
In other words, if a library has no priority record associated with it, it will 
not be used in scaffolding.<BR><BR>A priority can also be specified for a specific 
redundancy parameter. For example, the line:<BR><FONT FACE="Courier New, Courier, monospace">priority 
redundancy 4 1</FONT><BR>specifies that those links between contigs supported 
by 4 or more sources of linking data should be processed at priority level 1. 
Note, however, that you need to also specify a set of libraries to be used at 
this priority level. A simple shortcut is:<BR><FONT FACE="Courier New, Courier, monospace">priority 
ALL 1</FONT><BR>specifying that all link classes should be processed at this redundancy..<BR> 
<BR> If you are uncertain of which libraries your project uses, run Bambus without 
a configuration file. The libraries will be listed in a file ending in &quot;.libs&quot;.</P><LI> 
<P><B>Redundancy information.</B> Specifies the minimum number of links required 
for a valid inter-contig link. You can specify this parameter as a global value:<BR> 
<FONT FACE="Courier, monospace">redundancy 2</FONT><BR> that is, all contigs must 
be linked by at least two links. You can also specify a per-library value:<BR> 
<FONT FACE="Courier, monospace">redundancy MUMmer 1</FONT><BR> that is, a single 
MUMmer link is required to link two contigs.</P><LI> <P><B>Minimum scaffold size.</B> 
Specifies a scaffold size cutoff. For example:<BR> <FONT FACE="Courier, monospace">mingroupsize 
10000</FONT><BR> implies that only scaffolds that contain more than 10kbp are 
present in the output.</P><LI> <P><B>Link size error. </B>For some link types, 
determining the exact link length can only be done within a certain error. This 
parameter specifies, as a percentage, the estimated error in size determination. 
For example, since MUMmer links make an assumption of evolutionary closeness between 
two genomes, the error associated with them should be proportional to the estimated 
evolutionary distance (as nucleotide % similarity) between the two genomes. The 
syntax for the error parameter is:<BR> <FONT FACE="Courier, monospace">error MUMmer 
0.05</FONT><BR> for an error of 5%.</P><LI> <P><B>Overlapping contigs allowed.</B> 
For each library you can control the link validation process by specifying whether 
adjacent contigs are allowed to overlap or not. By default mate-pair links assume 
adjacent contigs do not overlap, while MUMmer links allow contigs to overlap. 
This is equivalent to the configuration parameters:<BR> <FONT FACE="Courier, monospace">overlaps 
lib_1 N<BR> overlaps MUMmer Y</FONT></P></UL><P><FONT FACE="Courier New, Courier, mono"><FONT FACE="Thorndale">An 
example is provided in </FONT><A HREF="Manual_files/bambus.conf.txt"><FONT FACE="Thorndale">data/sample_files/bambus.conf</FONT></A><FONT FACE="Thorndale">.</FONT></FONT></P><H3 CLASS="western"><A NAME="matesfile"></A><FONT COLOR="#ff0000"><I>4.2 
The .mates file</I></FONT></H3><P STYLE="font-style: normal"><SPAN STYLE="background: transparent">The 
.mates file provides two types of information: library data, and mate-pair relationships 
between reads. </SPAN> </P><P STYLE="font-style: normal"><SPAN STYLE="background: transparent">Library 
data can be described in two formats:</SPAN></P><P STYLE="font-style: normal"><SPAN STYLE="background: transparent"><FONT FACE="Courier, monospace">library 
&lt;name&gt; &lt;min_size&gt; &lt;max_size&gt;</FONT><BR> and<BR> <FONT FACE="Courier, monospace">library 
&lt;name&gt; &lt;min_size&gt; &lt;max_size&gt; &lt;regexp&gt;</FONT></SPAN></P><P STYLE="font-style: normal"><SPAN STYLE="background: transparent">Both 
formats require you to name each library and to provide a size range for the inserts 
belonging to it. The second format allows you to also provide a Perl regular expression 
that describes the naming convention for reads belonging to the library. The part 
of the sequence name that represents the library name must be placed within parantheses. 
As an example, the regular expression for TIGR sequences (where the library is 
specified by the first 4 characters) is:</SPAN></P><P STYLE="font-style: normal"><FONT FACE="Courier, monospace"><SPAN STYLE="background: transparent">(....).*</SPAN></FONT></P><P STYLE="font-style: normal"><SPAN STYLE="background: transparent">Mate-pair 
relationships can also be described in two ways:</SPAN></P><P STYLE="font-style: normal"><SPAN STYLE="background: transparent"><FONT FACE="Courier, monospace">pair 
&lt;regexp_forw&gt; &lt;regexp_rev&gt;</FONT><BR> or<BR> <FONT FACE="Courier, monospace">&lt;seq_forw&gt; 
&lt;seq_rev&gt; &lt;library_name&gt;</FONT></SPAN></P><P STYLE="font-style: normal"><SPAN STYLE="background: transparent">The 
first format requires two regular expressions corresponding to the forward and 
reverse mates of an insert. Just like the library record, the portion of the name 
corresponding to the insert name must be placed within parantheses. Two reads 
that match the two regular expressions, and have the exact same section matched 
within the parantheses will be considered mates. The library will be determined 
from the regular expression associated with a &quot;library&quot; record. As an 
example, at TIGR the first 7 characters represent the insert, then are followed 
by an optional T or P and the primer name (F/R). The corresponding regular expression 
is:</SPAN></P><P STYLE="font-style: normal"><FONT FACE="Courier, monospace"><SPAN STYLE="background: transparent">pair 
(.......)[TP]?F (.......)[TP]?R</SPAN></FONT></P><P STYLE="font-style: normal"><SPAN STYLE="background: transparent">The 
second format is simply a list of sequence name pairs each followed by the name 
of the library they belong to. The corresponding library records must be listed 
in the file before the pairing data.</SPAN></P><P STYLE="font-style: normal"><SPAN STYLE="background: transparent">Note 
that fields in the .mates file must be separated by TAB characters otherwise the 
program will report an error. </SPAN> </P><P STYLE="font-style: normal"><SPAN STYLE="background: transparent">An 
example of a .mates file using the phred/phrap naming convention is provided in 
<A HREF="Manual_files/bambus.mates.txt">data/sample_files/bambus.mates</A>.</SPAN></P><H3 CLASS="western"><A NAME="repfile"></A><FONT COLOR="#ff0000"><I>4.3 
Repeat files</I></FONT></H3><P>The repeat files used in screening linking data 
conform to the output standards of the <A HREF="ftp://ftp.tigr.org/pub/software/repeatFinder">repeatFinder</A> 
program. The file contains 5 TAB-delimited values: contig ID, repeat name, left 
and right coordinates within the contig, and contig class. Only the first 4 values 
are essential as far as Bambus is concerned. Here is an example of repeat records:</P><P><FONT FACE="Courier, monospace">AC009139.7.3 
RPT1A 14554 96820 1<BR> AC026498.3.2 RPT1B 124208 206475 1</FONT></P><H3 CLASS="western"><a name="EVXML"></a><FONT COLOR="#ff0000"><I>4.4 
.evidence.xml</I></FONT></H3><P>This provides an XML representation of the input 
file. Below is a description of the data presented in this file (see also an <A HREF="Manual_files/bambus.evidence.xml.txt">example</A>):</P><P>The 
overall flow of the XML file is:</P><P>&lt;EVIDENCE&gt;<BR>&nbsp;&nbsp;&lt;LIBRARY&gt;<BR>&nbsp;&nbsp;&nbsp;&nbsp;&lt;INSERT&gt;<BR>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&lt;SEQUENCE/&gt;<BR>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&lt;SEQUENCE/&gt;<BR>&nbsp;&nbsp;&nbsp;&nbsp;&lt;/INSERT&gt;<BR> 
&nbsp;&nbsp;&nbsp;&nbsp;... <BR>&nbsp;&nbsp;&lt;/LIBRARY&gt;<BR>&nbsp;&nbsp;...<BR>&nbsp;&nbsp;&lt;CONTIG&gt;<BR>&nbsp;&nbsp;&nbsp;&nbsp;&lt;SEQUENCE/&gt;<BR>&nbsp;&nbsp;&nbsp;&nbsp;...<BR>&nbsp;&nbsp;&lt;/CONTIG&gt;<BR>&nbsp;&nbsp;...<BR>&nbsp;&nbsp;&lt;LINK&gt;<BR>&nbsp;&nbsp;&nbsp;&nbsp;&lt;CONTIG/&gt;<BR>&nbsp;&nbsp;&nbsp;&nbsp;&lt;CONTIG/&gt;<BR>&nbsp;&nbsp;&lt;/LINK&gt;<BR>&nbsp;&nbsp;...<BR>&lt;/EVIDENCE&gt;</P><TABLE WIDTH="100%" BORDER="1"><TR><TD><P>&lt;EVIDENCE<BR> 
&nbsp;&nbsp;ID = &quot;1&quot;<BR>&nbsp;&nbsp;DATE = &quot;12/15/03&quot;<BR>&nbsp;&nbsp;PROJECT 
= &quot;MyProject&quot;<BR>&nbsp;&nbsp;PARAMETERS = &quot;&quot;<BR>&gt;</P></TD><TD>this 
is the parent tag - the whole document occurs within &lt;EVIDENCE&gt; and &lt;/EVIDENCE&gt;<BR>some 
identifier for the file<BR>date when file was created<BR>verbose description of 
the project<BR>parameters used in creating this file<BR></TD></TR><TR><TD><P>&lt;LIBRARY<BR>&nbsp;&nbsp;ID 
= &quot;lib_1&quot;<BR>&nbsp;&nbsp;NAME = &quot;short&quot;<BR>&nbsp;&nbsp;MIN 
= &quot;1200&quot;<BR>&nbsp;&nbsp;MAX = &quot;3500&quot;<BR>&gt;</P></TD><TD>for 
each shotgun library you need a separate library tag. <BR>some identifier for 
the library<BR>verbose name for the library<BR>minimum size of inserts in this 
library<BR>maximum size of inserts in this library<BR></TD></TR><TR><TD>&lt;INSERT<BR>&nbsp;&nbsp;ID=&quot;ins_1&quot;<BR>&nbsp;&nbsp;NAME=&quot;GALBZ92&quot;<BR>&gt;</TD><TD><P>the 
insert concept links together the two reads obtained from opposite ends of a clone 
insert<BR>generic identifier for the insert<BR>verbose name for the insert<BR></P></TD></TR><TR><TD>&lt;SEQUENCE<BR>&nbsp;&nbsp;ID=&quot;seq_1&quot;<BR>&nbsp;&nbsp;NAME=&quot;GALBZ92TF&quot;<BR>&gt;</TD><TD>each 
insert contains two sequence reads obtained from to the opposite ends<BR>generic 
identifier for the sequence<BR>verbose name for the sequence<BR></TD></TR><TR><TD>&lt;CONTIG<BR>&nbsp;&nbsp;ID=&quot;contig_1&quot;<BR>&nbsp;&nbsp;NAME= 
&quot;1&quot;<BR>&nbsp;&nbsp;LEN=&quot;12352&quot;<BR>&gt; </TD><TD><P>Each contig 
in the assembly needs to be represented in the file<BR>contig identifier<BR>contig 
name (usually the same as the identifier)<BR>contig length in basepairs<BR></P></TD></TR><TR><TD>&lt;SEQUENCE<BR>&nbsp;&nbsp;ID=&quot;seq_1&quot;<BR>&nbsp;&nbsp;ORI=&quot;BE&quot;<BR>&nbsp;&nbsp;ASM_LEND=&quot;0&quot;<BR>&nbsp;&nbsp;ASM_REND=&quot;525&quot;<BR>&gt;</TD><TD>each 
sequence in the contig is listed together with information about the position 
within the contig.<BR>sequence identifier (must match one described in the library 
section)<BR>orientation of sequence: BE - forward, EB - reverse<BR>coordinate 
of sequence's left end within the contig<BR>coordinate of sequence's right end 
within the contig<BR></TD></TR><TR><TD>&lt;LINK<BR>&nbsp;&nbsp;ID = &quot;link_1&quot;<BR>&nbsp;&nbsp;SIZE 
= &quot;-800&quot;<BR>&nbsp;&nbsp;TYPE = &quot;MUMmer&quot;<BR>&gt;</TD><TD><P>this 
is a generic link between two contigs (i.e. not inferred from mate-pair data)<BR>link 
identifier<BR>size of gap between the two contigs<BR>link type: links with the 
same type get grouped into a virtual &quot;library&quot; used in specifying priorities<BR></P></TD></TR><TR><TD>&lt;CONTIG<BR>&nbsp;&nbsp;ID 
= &quot;contig_1&quot;<BR>&nbsp;&nbsp;ORI = &quot;EB&quot;<BR>&gt;</TD><TD>each 
link contains two contigs<BR>contig identifier: must match one described in the 
contig section above<BR>contig orientation: BE - forward, EB - reverse<BR>each 
contig record may contain free-form data specifying the evidence for linking: 
e.g. alignment data</TD></TR></TABLE><P>&nbsp;</P><P>&nbsp;</P><H3 CLASS="western"><A NAME="outxml"></A>4.5 
.out.xml</H3><P>Represents, together with the evidence file, a description of 
the layout of the contigs. For each scaffold, the order (given as a coordinate 
along a chromosome) and the orientation of each contig is presented, together 
with a list of all the links used to generate this layout. Some links are deemed 
invalid, being given a code of &quot;LEN&quot; in case the length was deemed incorrect, 
or &quot;ORI&quot; in case the link orientation was considered incorrect. This 
file is also specified by the <A HREF="Manual_files/scaffold.dtd.txt">DTD</A> 
and an <A HREF="Manual_files/bambus.out.xml.txt">example (data/sample_files/bambus.out.xml)</A>. 
</P><H3 CLASS="western"><FONT COLOR="#ff0000"><I><a name="summary"></a>4.6 .stats 
file</I></FONT></H3><P>Finally, BAMBUS outputs a summary of the scaffolds generated. 
This file is pretty much self-explanatory, as evidenced by this <A HREF="Manual_files/bambus.stats.txt">example 
(data/sample_files/bambus.stats)</A>. Note that in the <FONT FACE="Courier New, Courier, monospace">.stats</FONT> 
file, the N50 sizes are computed with respect to the total span of the scaffolds 
unless a different genome size is specified in a file called <FONT FACE="Courier New, Courier, monospace">genome.size</FONT>.</P><H3 CLASS="western"><A NAME="details"></A>4.7 
.details file</H3><P>The .details file contains detailed information about the 
linking information between adjacent contigs. The contigs' orientation, size, 
and coordinates are listed and then all the linking data grouped by validity and 
library. An example is provided in <A HREF="Manual_files/bambus.details.txt">data/sample_files/bambus.details</A>.</P><H3 CLASS="western"><A NAME="dotfile"></A><FONT COLOR="#ff0000"><I>4.8 
.dot file</I></FONT></H3><P>Represents a graphical representation of the scaffolds 
in GraphViz format. Please see the <A HREF="http://www.research.att.com/sw/tools/graphviz/">AT&amp;T 
GraphViz</A> website for more information on this file format.</P><H3 CLASS="western"><A NAME="psfile"></A>4.9 
.ps file</H3><P>A postscript image generated from the GraphViz-formatted file. 
It can be obtained from the .dot file with the command:<BR> <FONT FACE="Courier, monospace">dot 
-Tps -o prefix.ps prefix.dot </FONT><BR> An example is given in the picture below.</P><P><BR> 
<BR> </P><H3 CLASS="western"><A HREF="Manual_files/display.jpg"><IMG SRC="Manual_files/display.jpg" NAME="Graphic2" ALIGN=LEFT WIDTH=860 HEIGHT=130 BORDER=0></A><BR CLEAR=LEFT> 
<BR> <BR> </H3><P>Each scaffold is placed in a box, labeled with some statistics 
on the scaffold size (number of contigs, number of bases and span). All the &quot;gaps&quot; 
- linking relationships between the contigs are represented as edges. Each edge 
is decorated with the number of links contributing to it, the number of links 
from each library type, and the number of links invalidated due to incorrect length 
(L) or orientation (O). Each contig's ID, size, and coordinates within the scaffold 
are also listed. </P><H3><A NAME="oo"></A>4.10 .oo file</H3><P>This file contains 
a summary of the order and orientation of all contigs present in the data. Each 
scaffold starts with a FASTA-like header containing the identifier for the scaffold 
followed by the number of contigs, size and span. Within each scaffold, all contigs 
are listed in the scaffold order, followed by the string BE for those in the forward 
orientation and EB for those in the reverse orientation. An example file is shown 
in: <A HREF="Manual_files/bambus.oo.txt">data/sample_files/bambus.oo</A>.</P><H3><A NAME="sum"></A>4.11 
.sum file</H3><P>This file contains one line for each scaffold. Each line contains 
the scaffold ID, number of contigs, scaffold size, and span. The four values are 
separated by TAB characters. See an example in <A HREF="Manual_files/bambus.sum.txt">data/sample_files/bambus.sum</A>.</P><H3 CLASS="western"><A NAME="problems"></A><I><FONT COLOR="#ff0000"><BR> 
5. Known problems</FONT></I></H3><SPAN STYLE="font-style: normal"><FONT COLOR="#000000">T</FONT></SPAN>here 
is a small &quot;off-by-one&quot; error in computing contig coordinates. It should 
not really affect the usefulness of the output. <H3 CLASS="western"><A NAME="contact"></A><FONT COLOR="#ff0000"><I><BR> 
6. Contact information</I></FONT></H3><P>BAMBUS is currently provided AS-IS, in 
other words we do not provide any support for the software. We would, however, 
like to hear your comments and suggestions. <A HREF="/tigr-scripts/mailer_scripts/contact_us.pl?receiver=bambus&amp;refer_text=BAMBUS">Click 
here to contact the BAMBUS team</A>.</P><P>BAMBUS was written by Mihai Pop and 
Dan Kosack.</P></TD></TR> </TABLE><P><BR><BR> </P>
</BODY>
</HTML>
